text/0102-offchain-parachain-runtime-upgrades.md

```diff
@@ -0,0 +1,354 @@
+expected to become more frequent and even attacks on service quality for other
+parachains would become a higher risk.
+
+## Stakeholders
+
+- Parachain Teams
+- Relay Chain Node implementation teams
+- Relay Chain runtime developers
+
+## Explanation
+
+The issues with on-chain runtime upgrades are:
+
+1. Needlessly costly.
+2. A single runtime upgrade more or less occupies an entire relay chain block, thus it
+   might affect also other parachains, especially if their candidates are also
+   not negligible due to messages for example or they want to uprade their
+   runtime at the same time.
+3. The signalling of the parachain to notify the relay chain of an upcoming
+   runtime upgrade already contains the upgrade. Therefore the only way to rate
+   limit upgrades is to drop an already distributed update in the size of
+   megabytes: With the result that the parachain missed a block and more
+   importantly it will try again with the very next block, until it finally
+   succeeds. If we imagine to reduce capacity of runtime upgrades to let's say 1
+   every 100 relay chain blocks, this results in lot's of wasted effort and lost
+   blocks.
+
+We discussed introducing a separate signalling before submitting the actual
+runtime, but I think we should just go one step further and make upgrades fully
+off-chain. Which also helps bringing down deposit costs in a secure way, as we
+are also actually reducing costs for the network.
+
+### Introduce a new UMP message type `RequestCodeUpgrade`
+
+As part of elastic scaling we are already planning to increase flexibility of [UMP
+messages](https://github.com/polkadot-fellows/RFCs/issues/92#issuecomment-2144538974), we can now use this to our advantage and introduce another UMP message:
+
+```rust
+enum UMPSignal {
+  // For elastic scaling
+  OnCore(CoreIndex),
+  // For off-chain upgrades
+  RequestCodeUpgrade(Hash),
+}
+```
+
+We could also make that new message a regular XCM, calling an extrinsic on the
+relay chain, but we will want to look into that message right after validation
+on the backers on the node side, making a straight forward semantic message more
+apt for the purpose.
+
+
+### Handle `RequestCodeUpgrade` on backers
+
+We will introduce a new request/response protocol for both collators and
+validators, with the following request/response:
+
+```rust
+struct RequestBlob {
+  blob_hash: Hash,
+}
+
+struct BlobResponse {
+  blob: Vec<u8>
+}
+```
+
+This protocol will be used by backers to request the PVF from collators in the
+following conditions:
+
+1. They received a collation sending `RequestCodeUpgrade`.
+2. They received a collation, but they don't yet have the code that was
+   previously registered on the relaychain. (E.g. disk pruned, new validator)
+
+In case they received the collation via PoV distribution instead of from the
+collator itself, they will use the exact same message to fetch from the valiator
+they got the PoV from.
```

Comment:
Why not make the code upgade simply be the parachain block?  Isn't that how substrate worked from the beginning?